### PR TITLE
 Laboratory work #2, Amina Abdeyeva - 23FPL1 

### PR DESCRIPTION
**Pull request by minabdeyeva**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -33,6 +60,32 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     """
+    clean_tokens = []
+
+    if not isinstance(tokens, list):
+        return None
+
+    if not isinstance(stopwords, list) or not stopwords:
+        return None
+
+    for word in stopwords:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:71]: use `any`

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +60,32 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+    for token in tokens:
+        if not isinstance(token, str):
+            clean_tokens = None
+            break
+        elif token in stopwords:
+            continue
+        clean_tokens.append(token)
+    if clean_tokens:
+        return clean_tokens
+    else:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:84]: you do not need `else`

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -33,6 +60,32 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+        return None
+
+    # remove_stopwords(['there', 'was', 'a', 'boy', 'named, 'henry'], ['was', 'named'])
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:87]: rm

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -18,6 +18,33 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    # divide by spaces - ['the', 'first%', 'sentence><.', 'the', 'sec&*ond', 'sent@ence', '#.']
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:23]: remove

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -18,6 +18,33 @@ def tokenize(text: str) -> list[str] | None:
...
+
+    text = text.lower()
+    text_by_space = text.split(' ')
+
+    separ_text = []
+    for word in text_by_space:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:29]: interesting implementation, but too dificult

**Comment by minabdeyeva** [comment]
changes are made, mistakes fixed

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -77,6 +164,38 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+
+        amount = 0
+        for doc in documents:
+            if voc in doc:
+                amount += 1
+
+        IDF = log((n - amount + 0.5) / (amount + 0.5))
+        map_IDF_scores[voc] = IDF
+
+    if map_IDF_scores:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:196]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -92,6 +211,38 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+    if not idf or not tf:
+        return None
+
+    map_scores = {}
+
+    for t in tf.keys():
+        score = tf[t] * idf[t]
+        map_scores[t] = score
+
+    if map_scores:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:242]: the same

**Comment by minabdeyeva** [comment]
done

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (minabdeyeva) [reply lab_2_retrieval_w_bm25/main.py:215]: fixed

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -18,6 +19,26 @@ def tokenize(text: str) -> list[str] | None:
...
+    text = text.replace('   ', ' ')
+    text = text.replace('  ', ' ')
+
+
+    text = text.strip()
+
+    tokens = text.split(' ')
+
+    if tokens == ['']:
+        tokens = []
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:39]: looks strange, please fine more elegant way

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -33,6 +54,27 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+    for token in tokens:
+        if not isinstance(token, str):
+            clean_tokens = None
+            break
+        if token in stopwords:
+            continue
+        clean_tokens.append(token)
+    if not clean_tokens:
+        return None
+    return clean_tokens
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:76]: try to return `clean_tokens or None`

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -63,6 +124,32 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
...
+            frequency[voc] = document_tokens.count(voc) / len(document_tokens)
+
+    for token in document_tokens:
+        if not isinstance(token, str):
+            return None
+        if token not in frequency.keys():
+            frequency[token] = document_tokens.count(token) / len(document_tokens)
+
+
+    if frequency:
```
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:142]: the same

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -77,6 +156,32 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+    for voc in vocab:
+        if not isinstance(voc, str):
+            return None
+        amount = 0
+        for doc in documents:
+            if voc in doc:
+                amount += 1
+        idf = log((len(documents) - amount + 0.5) / (amount + 0.5))
+        map_idf_scores[voc] = idf
+    if map_idf_scores:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:182]: the same

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (artyomtugaryov) [reply lab_2_retrieval_w_bm25/main.py:215]: no

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -121,6 +245,46 @@ def calculate_bm25(
...
+                    / (freq + k1 * (1 - b + b * (doc_len / avg_doc_len))))
+
+    for tok in document:
+        bm5_doc[tok] = 0.0
+        if tok in vocab:
+            freq = document.count(tok)
+            bm5_doc[tok] = (idf_document[tok] * freq * (k1 + 1)
+                    / (freq + k1 * (1 - b + b * ((abs(doc_len)) / avg_doc_len))))
+
+    if not bm5_doc:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:284]: the same

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -138,6 +302,45 @@ def rank_documents(
...
+                ranking_word += doc[word]
+        unsorted_ranking_map[i] = ranking_word
+
+    sorted_dict = dict(sorted(unsorted_ranking_map.items(), key=lambda item: item[1], reverse=True))
+    sorted_ranking_map = []
+
+    for sort_doc in sorted_dict:
+        sorted_ranking_map.append((sort_doc, sorted_dict[sort_doc]))
+
+    if not sorted_ranking_map:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:341]: the same

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by minabdeyeva** [comment]
done and fixed

**Pull request closed without merging**